### ISSUE TITLE
Automate Version Updates

### ISSUE THREAD

**Issue by andys8** [purescript/package-sets#704]
I've noticed quite often the package set is missing a recent version of a library. Manually checking it out, adding it, opening a PR, and waiting for the release, feels like it could be automated.

Asking @f-f directly I learned already a lot. Summarized here:

First of all `pursuit` or `pursuit-backups` is not the source of truth regarding library versions. Github is.

There is script that automatically bumped versions to the latest. It recently broke and needs to be fixed or rewritten:

https://github.com/spacchetti/pacchettibotti/blob/master/src/PacchettiBotti/Threads/PackageSets.hs

There will be a [new registry](https://discourse.purescript.org/t/rfc-a-new-registry-for-purescript-packages/1354). Once the registry is in place would be much easier. It will be backwards compatible to `package-sets`.

## Script

I think a basic script could look like this:

* Checkout `package-sets`
* For every package defined:
  * Look if there is a git tag with a newer version
  * Open question: What if dependencies changed?
  * If that's the case, create a pull request (if there is none already existing?)
  * CI on `package-sets` will check if the bump is okay

## Related Issues

See also https://github.com/purescript/package-sets/issues/6

**Comment by f-f** [comment]
Thanks for writing this down @andys8!

Indeed @pacchettibotti used to automatically open PRs to update packages, see e.g. https://github.com/purescript/package-sets/pull/633
However that code you linked broke some time ago and it got so involved so I'm now of the opinion that should be deprecated.
Most of the complexity in there stems from having to deal with "automatically opening PRs" (because then you have to consider all the edge cases of "is a PR already open", etc). I think just writing a script to run locally and then manually open a PR is a good way to start.
The script could just:
- query the latest release for every package in the set
- if the release we have in the set is different, put the new release in there
- run `make` to generate all the files, etc

About "what if dependencies change": once the registry is up we'll just get dependencies from the package manifest, right now we're duplicating the info here in package set because often the info in the bowerfile is inaccurate. So when dependencies changed in the past we just looked at the failing CI log and figured what needed to be tweaked. I.e. it has always been a manual process, and it's hard to automate.

**Comment by andys8** [comment]
Hey @f-f and @JordanMartinez 

I extended https://github.com/andys8/purescript-update-package-sets to update this issue https://github.com/purescript/package-sets/issues/728 . It will be closed and re-opened and its content will be kept up-to-date.

I hope this is a help to automatically check things.

Ideally it would also provide updates. Therefore the following non-trivial tasks are missing:

* parse dhall files
* update dhall files (they're split up grouped by author)
* generate `package.json` (relies on nix being available)
* create a git branch
* create a github pull request
* check if there is already an unmerged PR (to avoid spam)
* have a solution/plan in case a package update is blocked for longer time (due to failing ci)

**Comment by f-f** [comment]
Thanks for the great work so far @andys8! :heart: 

What's there is already of great help to keep things fresh. I have a few thoughts about your list for automating things further:
- since everything is running in GitHub Actions, getting Nix is as easy as adding one line to the workflow: https://github.com/purescript/package-sets/blob/169240863d681dc8c8c340584ece11caa0c44832/.github/workflows/main.yml#L11
- we don't need to parse/update Dhall directly from PureScript, as we can use `dhall-to-json` to get a JSON representation of every author file, manipulate that, and then use `json-to-dhall` to write the files back ([we already use it in the Registry CI](https://github.com/purescript/registry/blob/7b5534a141df1ef724e886ccf34d185016c5cdbe/ci/src/CheckRoundtrip.purs)). We already have these commands available in `nix-shell`, see [here](https://github.com/purescript/package-sets/blob/master/default.nix)
- in @pacchettibotti we kept a list of "banned versions" to avoid upgrading over and over to a version that wouldn't work. In there we had a database locally, but here we could just use a GitHub Issue for storing this list, as you're already doing for the list of packages that need to be updated
- would you be open to move that code to this repo? It might make logistics of running and mantaining it easier, as we could all have access to it if we need to take action in case the script goes crazy (this definitely happened to me with Pacchettibotti)

**Comment by andys8** [comment]
Thanks for the input and ideas 🙂

As discussed, once the current state (checking versions) has proven to be stable and working over same days, I plan to add it to this repository.

**Comment by thomashoneyman** [comment]
The [alpha for the new Registry](https://discourse.purescript.org/t/registry-alpha-launched/3146) has now launched, and this entails the automation of this repository - packages in the set will be managed through [the new API](https://github.com/purescript/registry#modify-the-package-set), and a new release will be cut once a day if there are any new updates.

I believe this solves this issue, but please open a new issue in the [`registry-dev` repo](https://github.com/purescript/registry-dev) if not.
